File: DataBinding/PageAndRDL/ODataDataSource/C#/JsonDataSourceClient/MainForm.cs
using GrapeCity.ActiveReports;
using System;
using System.IO;
using System.Windows.Forms;

namespace ODataDataSource
{
	/// <summary>
	/// Json Provider clinet, which get data via OData.
	/// </summary>
	public partial class MainForm: Form
    {
		private PageReport _pageReport;
		private const string _path = @"..\..\..\..\Reports\testReport.rdlx";
		private Service _service = Service.None;

		public MainForm()
        {
            InitializeComponent();
			var rptPath = new FileInfo(_path);
			_pageReport = new PageReport(rptPath);
			_pageReport.Document.LocateDataSource += OnLocateDataSource;
			comboBox1.SelectedIndex = 0;
        }

		// The handler of <see cref="PageDocument.LocateDataSource"/> that returns appropriate data for a report.
		private void OnLocateDataSource(object sender, LocateDataSourceEventArgs args)
		{
			object data = null;
			var dataSourceName = args.DataSet.Query.DataSourceName;
			if (dataSourceName == "DataSource1")
			{
				try
				{
					data = DataLayer.CreateData(_service);
				}
				catch (Exception ex)
				{
					MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
					return;
				}
			}
			args.Data = data;
		}

		private void comboBox1_SelectedIndexChanged(object sender, EventArgs e)
		{
			if ((int)_service == comboBox1.SelectedIndex)
				return;

			_service = (Service)Enum.Parse(typeof(Service), comboBox1.SelectedIndex.ToString());
			reportPreview.LoadDocument(_pageReport.Document);
		}
	}
}

File: DataBinding/PageAndRDL/ODataDataSource/VB.NET/JsonDataSourceClient/MainForm.vb
Imports System.IO
Imports GrapeCity.ActiveReports

''' <summary>
''' Json Provider clinet, which get data via OData.
''' </summary>
Public Class MainForm

	Private _pageReport As PageReport
	Private Const _path As String = "..\..\..\..\Reports\testReport.rdlx"
	Private _service As Service = Service.None

	Public Sub New()
		InitializeComponent()
		Dim rptPath As FileInfo = New FileInfo(_path)
		_pageReport = New PageReport(rptPath)
		AddHandler _pageReport.Document.LocateDataSource, AddressOf OnLocateDataSource
		comboBox1.SelectedIndex = 0
	End Sub


	' The handler of <see cref="PageDocument.LocateDataSource"/> that returns appropriate data for a report.
	Private Sub OnLocateDataSource(sender As Object, args As LocateDataSourceEventArgs)
		Dim data As String = Nothing
		Dim dataSourceName As String = args.DataSet.Query.DataSourceName
		If dataSourceName = "DataSource1" Then
			Try
				data = DataLayer.CreateData(_service)
			Catch ex As Exception
				MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
				Return
			End Try
		End If
		args.Data = data
	End Sub


	Private Sub comboBox1_SelectedIndexChanged(sender As Object, e As EventArgs) Handles comboBox1.SelectedIndexChanged
		If _service = comboBox1.SelectedIndex Then
			Return
		End If

		_service = CType([Enum].Parse(GetType(Service), comboBox1.SelectedIndex.ToString()), Service)
		reportPreview.LoadDocument(_pageReport.Document)
	End Sub
End Class
